Add support for the delimiter parameter to the List Objects API

hennejg opened the pull request:
## Description
Add support for `delimiter` parameter to the ListObjects and ListObjectsV2 APIs

## Related Issue
#89 

## Tasks
- [x] I have signed the [CLA](http://adobe.github.io/cla.html).
- [x] I have written tests and verified that they fail without my change.

agudian commented:
Looks good! Interesting feature! 

Thanks for controbuting it right away 👍

I see you‘re still working on it, let us know when you want us to start a review...

hennejg commented:
> I see you‘re still working on it, let us know when you want us to start a review...
I didn't find the test cases right away, since the test class is named `AmazonClientUploadIT` but it also contains tests unrelated to uploads. So, now there's a test case as well. And checkstyle is happy, too. 
I think we're good to go.

hennejg commented:
Thanks for the quick feedback!

Rebase/squash is done. In addition, I got rid of two spurious diffs that were due to my overzealous eclipse cleanup actions.

hennejg commented:
Unfortunately, the semantics of the `prefix` and `delimiter` parameters - especially if used in conjunction - are not described thoroughly in the AWS specicifation. I therefore mapped them systematically, and doing so, I noticed that there is still some drastic discrepancies between AWS and the the mock implementation. 

Please refrain from merging this PR to give me some more time to improve the mock's fidelity.

hennejg commented:
As promised, I've updated the PR with a few more changes to bring the mocked behaviour in line with AWS. 

I moved the relevant test case into a separate class, `ListObjectID`, which allows me to easily cover many cases using a parameterized test.

agudian commented on server/src/test/java/com/adobe/testing/s3mock/its/AmazonClientUploadIT.java:
```diff
@@ -512,6 +514,8 @@ public void shouldDeleteBucket() {
   /**
    * Tests if the list objects can be retrieved.
+   * 
+   * <p>For more detailed tests of the List Objects API 
```
Is this sentence cut off?

hennejg replied:
Yep, indeed. Fixed.

agudian commented on server/src/main/java/com/adobe/testing/s3mock/domain/FileStore.java:
```diff
@@ -502,11 +503,22 @@ public S3Object getS3Object(final String bucketName, final String objectName) {
+    
+    // Determine whether the prefix ends with a path separator by looking at
+    // what adding some definitely non-separator stuff does to equality.
+    boolean endsWithSeparator = !Strings.isEmpty(prefix)
```
don't we always have `/` as path separator here? If so, we could just use endsWith on `prefix`?

hennejg replied:
No, this is not the case, and it illustrates what I think is another deficiency of the `FileStore` I meant to  mention, but forgot about: the S3 API lets you use just about any character (or sequence thereof) as a delimiter, so you could use e.g. a colon as a hierarchy separator. This is due to the fact that by and large, S3 treats object keys as flat Strings, not a hierarchy of folders. The hierarchy is simulated during queries with arbitrary prefixes and delimiters.
However, if you actually use a delimiter other than `/`, I suspect that FileStore will break in a few places.

This weird looking code deals with the question whether the supplied prefix ends with whatever the `Path` API accepts as a path sepatator.

IMHO the whole FileStore could be refactored into a flat list of files (most tests will use a very limited number of objects I suppose) by appropriately escaping all keys. This would make things a lot easier, I guess.

agudian replied:
Very good point, I'll add an issue for that: #94.

hennejg replied:
Thanks. I just verified my suspicion with a test case using colons as separator, which indeed failed.

agudian commented on server/src/test/java/com/adobe/testing/s3mock/its/ListObjectIT.java:
```diff
@@ -0,0 +1,194 @@
+    );
+  }
+
+  private static final List<String> param = new ArrayList<>();
```
This looks like debug stuff... can we remove that?

hennejg replied:
This list was indeed used during development to collect the answers by AWS in order to build the expectations. I kept it in, because it might come in handy if someone wants to add more cases, but I can simple keep my development version around and get rid of it here.

agudian commented on server/src/test/java/com/adobe/testing/s3mock/its/ListObjectIT.java:
```diff
@@ -0,0 +1,194 @@
+  @Parameters(name = "{index}: prefix={0}, delimiter={1}")
+  public static Iterable<Object[]> data() {
+    return Arrays.<Object[]>asList(//
+        arr(null, null, ALL_OBJECTS, arr()), //
```
Using a dedicated data structure instead of four individual properties (with two of them being arrays again) could make this more readable. Maybe one with a fluent interface?

hennejg replied:
Yes, you are right about that. I'll see what I can do.

agudian commented on server/src/test/java/com/adobe/testing/s3mock/its/ListObjectIT.java:
```diff
@@ -0,0 +1,194 @@
+  public Object[] expectedPrefixes;
+
+
+  private static S3MockApplication s3mock;
```
The S3Mock is already expected to run at this point (as Docker container in the build or started manually in the IDE with default settings). I'd prefer we keep it with the same expectations here as well.

hennejg replied:
In fact, I was wondering about why you would opt for launching a separate container as opposed to just doing what this library is well suited to do: run it in embedded mode. Not having to deal with containers or manually started processes makes things so much easier.

agudian replied:
The ITs in this Maven module are supposed to verify the stand-alone app, and that's done through the Docker image here.

Testing the embedded mode is primarily done through the various `testsupport` modules.

I agree with the manual starting of the other process being a bit inconvenient... it's a trade-off, but in the end I find it a bit more important to have consistency in the ITs of a single module.

agudian commented on server/src/test/java/com/adobe/testing/s3mock/its/ListObjectIT.java:
```diff
@@ -0,0 +1,194 @@
+  public static void createBucket() {
+    s3mock = S3MockApplication.start();
+
+    s3client = RUN_AGAINST_AWS ? createAwsClient() : createMockClient();
```
As running against AWS might not be very well portable for other contributors, so I'd like to avoid having such dead code paths here... I understand that it was helpful while developing, though... :)

How about extending `S3TestBase` and using the s3Client from there?

hennejg commented:
I have updated my PR with the changes as per your review. Please let me know if you like them.

agudian commented:
Alright, thanks a lot!
